### DISCUSSION TITLE
Further cursor coloring enhancement

### DISCUSSION THREAD

**Discussion by VKondakoff** [kovidgoyal/kitty#4136] [Ideas]
Hi!

If you use fixed cursor colors you may experience something like this (cursor is invisible on the same background):

![ezgif com-gif-maker](https://user-images.githubusercontent.com/230824/137765336-cb0cea3d-214b-41c0-80bf-0dc2702095f6.gif)

There is an useful option 'cursor_text_color background' to enhance the issue. Here is how it looks in my example:

![ezgif com-gif-maker-2](https://user-images.githubusercontent.com/230824/137765533-b2e44eee-009d-462a-8455-0e64682ac7b9.gif)

It is better: you can see where the cursor is, but still not acceptable, because the character inside cursor is not visible.
Can we, please, made another one step to enhance cursor coloring. I'm suggesting two things:
1. Set cursor background to a color of a character under cursor
2. Set character inside cursor color to a cell background color

As far as I see this way we will always see the cursor and the character inside it.
What do you think?

**Comment by kovidgoyal** [comment]
I already said what I think on the issue about it.

**Reply by VKondakoff** [reply]
Can you, please, provide a link? I'm a bin new here.
Thanks!

**Reply by kovidgoyal** [reply]
You posted on that issue already.

**Comment by Luflosi** [comment]
https://github.com/kovidgoyal/kitty/issues/126

**Comment by VKondakoff** [comment]
Aha, thanks!
Are we talking about this?

> There are three possible approaches:
> Use a defined foreground and background color
> Use a defined background color and the foreground color from the current terminal background
> Use a full reverse video effect (invert the colors in the terminal for the cursor)
> Different terminals typically support one of these three approaches. And different kitty users have requested all three at different times. I am fine with (1) and (2). The problem with (3) is that it breaks the terminal escape codes used to change cursor colors.
> That said I also dont much see the point of (1). You should set your cursor color to have good contrast with the background color. If you do that and you still have poor contrast with text , then that means the text already has poor contrast with the background, and is hard to see regardless of whether the cursor is over it or not.

If the answer is 'yes', I just wanted to note - I'm proposing the fourth approach (4): dynamically swapping the foreground and background colors to minimize the chance of 'unreadable' cursor. Or this is considered the same as 'full reverse video effect' (3)? (Sorry, this terminology is quite new to me and I'm not into the escape codes yet - so I'm not sure if this will break the terminal escape codes used to change cursor colors or not).

**Comment by kovidgoyal** [comment]
On Mon, Oct 18, 2021 at 09:47:41AM -0700, VKondakoff wrote:
> Aha, thanks!
> Are we talking about this?
> 
> > There are three possible approaches:
> > Use a defined foreground and background color
> > Use a defined background color and the foreground color from the current terminal background
> > Use a full reverse video effect (invert the colors in the terminal for the cursor)
> > Different terminals typically support one of these three approaches. And different kitty users have requested all three at different times. I am fine with (1) and (2). The problem with (3) is that it breaks the terminal escape codes used to change cursor colors.
> > That said I also dont much see the point of (1). You should set your cursor color to have good contrast with the background color. If you do that and you still have poor contrast with text , then that means the text already has poor contrast with the background, and is hard to see regardless of whether the cursor is over it or not.
> 
> If the answer is 'yes', I just wanted to note - I'm proposing the fourth approach (4): dynamically swapping the foreground and background colors to minimize the chance of 'unreadable' cursor. Or this is considered the same as 'full reverse video effect' (3)? (Sorry, this terminology is quite new to me and I'm not into the escape codes yet - so I'm not sure if this will break the terminal escape codes used to change cursor colors or not).

Yes its the same as 3 and yes it will break the escape codes.
